Sources/WebSocketKit/PMCE.swift

```diff
@@ -0,0 +1,753 @@
+                                                 eventLoop: channel.eventLoop)
+            
+            self.decompressorBox = NIOLoopBoundBox( CompressionAlgorithm.deflate(configuration: zcdConf).decompressor,
+                                             eventLoop: channel.eventLoop)
+            
+            logger.trace("compressor \(zccConf)")
+            logger.trace("decompressor \(zcdConf)")
+        }
+        startStreams()
+    }
+    
+    ///  Starts compress-nio streams for DEFLATE support.
+    /// - returns: Void
+    public func startStreams() {
+        do {
+            try compressorBox.value?.startStream()
+        }
+        catch {
+            logger.error("error starting compressor stream : \(error)")
+        }
+        do {
+            try decompressorBox.value?.startStream()
+        }
+        catch {
+            logger.error("error starting decompressor stream : \(error)")
+        }
+    }
+    
+    ///  Stops compress-nio streams for DEFLATE support.
+    /// - returns: Void
+    public func stopStreams() {
+        do {
+            logger.debug("PMCE: stopping compressor stream...")
+            try compressorBox.value?.finishStream()
+        }
+        catch {
+            logger.error("PMCE:error finishing stream(s) : \(error)")
+        }
+        
+        do {
+            logger.debug("PMCE: stopping decompressor stream...")
+            try decompressorBox.value?.finishStream()
+        }
+        catch {
+            logger.error("PMCE:error finishing stream(s) : \(error)")
+        }
+        
+
+    }
+
+    /// websocket send calls this to compress.
+    public func compressed(_ buffer: ByteBuffer,
+                            fin: Bool = true,
+                            opCode: WebSocketOpcode = .binary) throws -> WebSocketFrame {
+        
+        guard let channel = channel else {
+            throw IOError(errnoCode: 0, reason: "PMCE: channel not configured.")
+        }
+        let notakeover = !shouldTakeOverContext()
+
+        let startSize = buffer.readableBytes
+        
+        if logging {
+            logger.trace("compressing \(startSize) bytes for \(opCode)")
+        }
+        do {
+            var mutBuffer = buffer
+
+            if !notakeover {
+                mutBuffer = unpad(buffer:buffer)
+            }
+
+            let startTime = Date()
+            
+            let compressed =
+            try mutBuffer.compressStream(with: compressorBox.value!,
+                                         flush: .sync,
+                                         allocator: channel.allocator)
+                                        //rfc says to strip padding word off compressed
+            mutBuffer = unpad(buffer: mutBuffer)
+            if logging {
+                let endTime = Date()
+                let endSize = compressed.readableBytes
+                
+                logger.trace("compressed \(startSize) to \(endSize) bytes @ \(startSize / endSize) ratio from")
```

Comment:
Doing the ratio and time calculation on every frame is expensive and clutters the code. For instrumenting at this level, prefer a pair of properties on `self` such as `uncompressedBytes` and `compressedBytes` which have the pre- and post-compression sizes added to them in this method; the overall ratio can then be derived from them as needed, and the time cost can be derived from performance benchmarking at a (much) higher level.

(Oh, and do _NOT_ use an `NIOLockedValueBox` on the stats properties, they're already guarded by their limited usage.)

Reply:
anything not a let constant has to be sendable and thus boxed though wouldnt it? If I recall mutability was its bane.

Reply:
That was what I was told to do when sendable came up unexpected. I dont pretend to understand all this.

Reply:
Ive removed the calculations as they were more there for my client's stated goals and since he hasnt been bothered to squeak in a few months now. SO he could get an idea if the settings changes where achieving what he wanted.
It doesnt really apply to the general use case I think.